File: Python/FactoryMethod/FactoryMethod.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
Factory Method Pattern
Author: Kei Nakata
Data: Oct.7.2014
'''
import abc

class Creator(object):
    __meataclass__ = abc.ABCMeta

    @abc.abstractmethod
    def create(self):
        pass

    @abc.abstractmethod
    def factoryMethod(self):
        pass

class Product(object):
    __meataclass__ = abc.ABCMeta
    @abc.abstractmethod
    def call(self):
        pass

class ConcreteCreator(Creator):
    def __init__(self):
        self.count = 0

    def create(self):
        self.count = self.count + 1
        return ConcreteProduct()
    def factoryMethod(self):
        print "I spawned", self.count, "times!"

class ConcreteProduct(object):
    def call(self):
        print "Hi!"

if __name__ == '__main__':
    A = ConcreteCreator()
    a = A.create();
    b = A.create();
    c = A.create();
    d = A.create();
    a.call()
    b.call()
    c.call()
    d.call()
    A.factoryMethod()

File: C++/FactoryMethod/FactoryMethod.cpp
/*
 * Prototype Pattern
 * Author: Kei Nakata
 * Data: Oct.6.2014
 *
 */
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <stack>
using std::cout;
using std::endl;
using std::string;
using std::stack;

class Product
{
public:
    virtual void call() = 0;
    virtual ~Product(){}
};

class Creator
{
public:
    virtual Product* create() = 0;
    virtual void factoryMethod() = 0;
    virtual ~Creator() {}
};

class ConcreteProduct: public Product
{
public:
  void call()
  {
    cout << "Hi!" << endl;
  }
};

template<typename T>
class LifeTimeManager {
    stack<T*> store;
public:
    void push(T* p)
    {
        store.push(p);
    }
    ~LifeTimeManager()
    {
        while(!store.empty())
        {
            if(store.top() != NULL)
                delete store.top();
            store.pop();
        }
    }
};

class ConcreteCreator : public Creator
{
  int count;
  LifeTimeManager<Product> manager;
public:
    ConcreteCreator() : count(0)
    {}

    Product* create()
    {
      this->count++;

      Product* p = new ConcreteProduct;
      manager.push(p);
      return p;
    }

    void factoryMethod()
    {
      cout << "I spawned " << this->count << " times!" << endl;
    }
};

int main()
{
  ConcreteCreator mommy;
  Product *a, *b, *c, *d;
  a = mommy.create();
  a->call();
  b = mommy.create();
  b->call();
  c = mommy.create();
  c->call();
  d = mommy.create();
  d->call();
  mommy.factoryMethod();
  return EXIT_SUCCESS;
}
